plutus-conformance/src/PlutusConformance/Common.hs

```diff
@@ -104,20 +39,147 @@ shownParseError = "parse error"
 type UplcProg = UPLC.Program Name DefaultUni DefaultFun ()
 
+-- UPLC evaluation test functions
+
+-- | The evaluator to be tested. It should either return a program if the evaluation is successful,
+-- or `Nothing` if not.
 type UplcEvaluator = UplcProg -> Maybe UplcProg
 
--- | A UPLC evaluation test suite.
-data UplcEvaluationTest =
-    MkUplcEvaluationTest {
-       -- | The evaluator function to use for the test.
-       evaluator :: UplcEvaluator
-       -- | The test directory in which the test files are located.
-       , testDir :: FilePath
-    }
+-- | Walk a file tree, making test groups for directories with subdirectories,
+-- and test cases for directories without.
+discoverTests :: UplcEvaluator -- ^ The evaluator to be tested.
+    -> (FilePath -> Bool)
+    -- ^ A function that takes a test name and returns
+    -- whether it should labelled as `ExpectedFailure`.
+    -> FilePath -- ^ The directory to search for tests.
+    -> IO TestTree
+discoverTests eval expectedFailureFn dir = do
+    let name = takeBaseName dir
+    children <- listDirectory dir
+    subdirs <- flip wither children $ \child -> do
+        let fullPath = dir </> child
+        isDir <- doesDirectoryExist fullPath
+        pure $ if isDir then Just fullPath else Nothing
+    if null subdirs
+    -- no children, this is a test case directory
+    then do
+            let goldenFilePath = dir </> name <> ".uplc.expected"
+                mkGoldenTest =
+                        goldenTest
+                            name -- test name
+                            -- get the golden test value
+                            (expectedToProg <$> T.readFile goldenFilePath)
+                            -- get the tested value
+                            (getTestedValue eval dir)
+                            (\ x y -> pure $ compareAlphaEq x y) -- comparison function
+                            (updateGoldenFile goldenFilePath) -- update the golden file
+            -- if the test is expected to fail, mark it so.
+            if expectedFailureFn dir
+            then pure $ expectFail mkGoldenTest
+            -- the test isn't expected to fail, make the `TestTree` as usual.
+            else pure mkGoldenTest
+    -- has children, so it's a grouping directory
+    else testGroup name <$> traverse (discoverTests eval expectedFailureFn) subdirs
+
+-- | Turn the expected file content in text to a `UplcProg` in Debruijn unless the expected result
+-- is a parse or evaluation error.
+expectedToProg :: T.Text -> Either T.Text UplcProg
+expectedToProg txt
+  | txt == shownEvaluationFailure =
+    Left txt
+  | txt == shownParseError =
+    Left txt
+  | otherwise =
+    case parseTxt txt of
+        Left _     -> Left txt
+        Right p -> -- The evaluator throws if the term has free variables
+            case UPLC.deBruijnTerm (UPLC._progTerm (void p)) of
+                Left (_ :: UPLC.FreeVariableError) -> Left shownEvaluationFailure
```

Comment:
I don't think there should be any case where the _expected_ behavior is that the evaluator returns a `Right uplcProg`, but the `uplcProg` cannot be deBruined, are there?

Reply:
Great question! Is it maybe possible to return a lambda from the evaluator and sneak a free type variable in there?

Reply:
I was thinking of just removing the `UPLC.deBruijnTerm`. The expected `p` should never fail to be deBruijnized. If somehow the actual `p` does fail to be deBruijnized, the test should just fail.

Reply:
I think this makes things worse. If the expected program, for some reason, fails to deBruijnize, and the actual program fails to evaluate, then both will be `Left shownEvaluationFailure` and your test will succeed, while it should fail. So I'd recommend removing this, and just have `Right p -> Right $ void p`.

Reply:
actually I just realized that this won't be evaluated and so there's no need to deBruijnize and catch the exception here. Thanks for spotting this!

Reply:
OK so this was code when I thought we need to deBruijnize, so I've removed that in the `getTestedValue` function as well.